sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -46,6 +46,21 @@ trait CheckAnalysis extends PredicateHelper {
     }).length > 1
   }
 
+  private def checkLimitClause(limitExpr: Expression): Unit = {
+    if (!limitExpr.foldable) {
+      failAnalysis(
+        "The argument to the LIMIT clause must evaluate to a constant value. " +
+        s"Limit:${limitExpr.sql}")
+    }
+    limitExpr.eval() match {
+      case o: Int if o >= 0 => // OK
+      case o: Int => failAnalysis(
+        s"number_rows in limit clause must be equal to or greater than 0. number_rows:$o")
+      case o => failAnalysis(
```

Comment:
Tried to do it in different ways, but the best way I found is to introduce an Analyzer rule. 

We have one rule in Optimizer and one rule in Planner. Both assume the `limitExpr` is an `Integer` type. It looks ugly to convert the type everywhere. To support different numeric types, we need to convert these foldable numeric values to integer in `Analzyer`, I think. Let me upload a fix at first. You can judge whether this is a good way or not. Thanks!